src/compiler/scala/tools/nsc/transform/CleanUp.scala

```diff
@@ -383,6 +380,106 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
       }
     }
 
+    object StringsPattern {
+      def unapply(arg: Tree): Option[List[String]] = arg match {
+        case Literal(Constant(value: String)) => Some(value :: Nil)
+        case Literal(Constant(null))          => Some(null :: Nil)
+        case Alternative(alts)                => traverseOpt(alts)(unapply).map(_.flatten)
+        case _                                => None
+      }
+    }
+
+    // transform scrutinee of all matches to ints
+    def transformSwitch(sw: Match): Tree = { import CODE._
+      sw.selector.tpe match {
+        case IntTpe => sw // can switch directly on ints
+        case StringTpe =>
+          // these assumptions about the shape of the tree are justified by the codegen in MatchOptimization
+          val Match(Typed(selTree: Ident, _), cases) = sw
+          val sel = selTree.symbol
+          val restpe = sw.tpe
+          val swPos = sw.pos.focus
+
+          /* From this:
+           *     string match { case "AaAa" => 1 case "BBBB" | "c" => 2 case _ => 3}
+           * Generate this:
+           *     goto matchSuccess (string.## match {
```

Comment:
Why "goto matchSuccess" here? All cases of the match branch so it seems like a no-op to me.

Reply:
Nice catch. I had it before I realized the problem with `Alternative`s